letsencrypt-apache/letsencrypt_apache/configurator.py

```diff
@@ -731,6 +732,67 @@ def enhance(self, domain, enhancement, options=None):
             logger.warn("Failed %s for %s", enhancement, domain)
             raise
 
+    def _set_http_header(self, ssl_vhost, header_name):
+        """Enables header header_name on ssl_vhost.
+
+        If header_name is not already set, a new Header directive is placed in
+        ssl_vhost's configuration with arguments from:
+        constants.HTTP_HEADER[header_name]
+
+        .. note:: This function saves the configuration
+
+        :param ssl_vhost: Destination of traffic, an ssl enabled vhost
+        :type ssl_vhost: :class:`~letsencrypt_apache.obj.VirtualHost`
+
+        :param header_name: a header name, e.g: Strict-Transport-Security
+        :type str
+
+        :returns: Success, general_vhost (HTTP vhost)
+        :rtype: (bool, :class:`~letsencrypt_apache.obj.VirtualHost`)
+
+        :raises .errors.PluginError: If no viable HTTP host can be created or
+            set with header header_name.
+
+        """
+        if "headers_module" not in self.parser.modules:
+            self.enable_mod("headers")
+
+        # Check if selected header is already set
+        self._verify_no_http_header(ssl_vhost, header_name)
+
+        # Add directives to server
+        self.parser.add_dir(ssl_vhost.path, "Header",
+                constants.HEADER_ARGS[header_name])
+
+        self.save_notes += ("Adding %s header to ssl vhost in %s\n" %
+                (header_name, ssl_vhost.filep))
+
+        self.save()
+        logger.info("Adding %s header to ssl vhost in %s", header_name,
+                ssl_vhost.filep)
+
+    def _verify_no_http_header(self, ssl_vhost, header_name):
+        """Checks to see if existing header_name header is in place.
+
+        Checks to see if virtualhost already contains a header_name header
+
+        :param vhost: vhost to check
+        :type vhost: :class:`~letsencrypt_apache.obj.VirtualHost`
+
+        :returns: boolean
+        :rtype: (bool)
+
+        :raises errors.PluginError: When header header_name exists
+
+        """
+        header_path = self.parser.find_dir("Header", None, start=ssl_vhost.path)
+        if header_path:
+            # "Existing Header directive for virtualhost"
+            for match in header_path:
+                if self.aug.get(match).lower() == header_name.lower():
```

Comment:
The use of equality as the test here is slightly dangerous.  For instance, if I have an vhost entry like this:
```
Header add "Content-Security-Policy" "upgrade-insecure-requests"
```
then this test will return false, because it's comparing the strings `"upgrade-insecure-requests"` and `upgrade-insecure-requests`.  So you wind up with this in your conf file after `letsencrypt install --uir`:

```
Header add "Content-Security-Policy" "upgrade-insecure-requests"
Header always set Content-Security-Policy upgrade-insecure-requests
```
I don't know what a more correct alternative is, though.  Substring matching doesn't work either because there could be something like this in the conf file:
```
Header append Content-Security-Policy "script-src 'self' upgrade-insecure-requests.info"
``` 
We don't want to have to write parsers for every type of header we support, or at least not yet, so some zen philosophy might be required in working around this issue.